webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DataStructures.scala

```diff
@@ -579,58 +579,114 @@ final case class BoundPolicyDraft(
+                                             s"In '${directiveName}' [${id.directiveId.debugString}] in rule '${ruleName}' [${id.ruleId.serialize}]: component Id are " +
+                                             s"not the same for technique and var for '${cid.value}'. In technique: '${cid2.parents.reverse
+                                                 .mkString("/")}' ; In var: '${cid.parents.reverse.mkString("/")}'"
+                                           )
+                                         }
+                                         // check if there is mandatory variable with missing/blank values
+                                         (
+                                           variable.spec.constraint.mayBeEmpty,
+                                           variable.values.exists(v => v == null || v.isEmpty)
+                                         ) match {
+                                           // simple case: it's optional, we don't care if empty or not.
+                                           case (true, _) => Right((cid, variable))
+
+                                           // simple case: all mandatory are filled
+                                           case (false, false) => Right((cid, variable))
+
+                                           // complicate case: mandatory with unfilled value: check if a default is available
+                                           case (false, true) =>
+                                             // first, if we have a default value, use that
+                                             variable.spec.constraint.default match {
+                                               case Some(d) =>
+                                                 variable
+                                                   .copyWithSavedValues(variable.values.map(v => if (v.isEmpty) d else v))
+                                                   .map(v => (cid, v))
+                                               // else it's an error
+                                               case None    =>
+                                                 Left(
+                                                   Inconsistency(
+                                                     s"Error for policy for directive '${directiveName}' [${id.directiveId.debugString}] in rule '${ruleName}' [${id.ruleId.serialize}]: " +
+                                                     s"a non optional value is missing for parameter '${variable.spec.description}' [param ID: ${variable.spec.name}]"
+                                                   )
+                                                 )
+                                             }
+
+                                         }
+                                     }
+                                 }
+        missing                = techVarSpecs -- checkedExistingValues.map(_._1.value).toSet
+        // add missing if possible
```

Comment:
here, we check that the variables that are declared in the technique spec but didn't have a value in the draft are either optional or have an usable default value. And we add them, so that all variable defined in the technique exists (the optional just don't have any value).